### PR TITLE
custom_vjp bug

### PR DESCRIPTION
**Pull request by zou3519**
There's a bug in simple functorch's custom_vjp. In particular, we would
like to learn what it would take to get simple functorch's custom_vjp to
work like JAX's custom_vjp w.r.t. to the behavior towards intermediate
Tensors.

TODO: we should try to fix the bugs mentioned.

### CODE REVIEW

File: simple_functorch.py [hunk 1]
```diff
@@ -1277,3 +1277,71 @@ def run_gradvmap(d2: "Batched", d1: "Autograd"):
...
+def f_bwd(saved, grad_output):
+    two_x, = saved
+    return (grad_output * two_x,)
+
+jax_square.defvjp(f_fwd, f_bwd)
+
+# This is 2.0 which is correct
+ddx = jax.grad(jax.grad(jax_square))(x)
+print(ddx)
+"""
```
Comment (zou3519) [inline simple_functorch.py:1346]: This is the "reference". Note that JAX's f_fwd returns an intermediate, two_x and uses it in the gradient computation. Computing the second order gradient is correct.

File: simple_functorch.py [hunk 2]
```diff
@@ -1277,3 +1277,71 @@ def run_gradvmap(d2: "Batched", d1: "Autograd"):
...
+    # [TapeEntry(inputs=['v197'], outputs=['v199'], propagate=<function Autograd.custom_vjp.<locals>
+    # .propagate at 0x7f8310658a60>), TapeEntry(inputs=['v200', 'v198'], outputs=['v201'], propagate
+    # =<function Autograd.mul.<locals>.propagate at 0x7f83106a3dc0>)]
+    return ddx
+
+# Bug 0:
+# This is wrong: The result is None, but it should be 2.
+# Somehow the gradients aren't getting recorded.
+ddx = run_gradgrad(outer, inner, x)
+import pdb; pdb.set_trace()
```
Comment (zou3519) [inline simple_functorch.py:1320]: Bug 0: ddx should equal two, but it is actually None here. This behavior is in-line with with autograd.Function does today, actually. autograd.Function requires the user to specify a gradient formula for the intermediate (two_x).

However, we're trying to explore what it would take to not require the user to specify the gradient formula.

File: simple_functorch.py [hunk 3]
```diff
@@ -1277,3 +1277,71 @@ def run_gradvmap(d2: "Batched", d1: "Autograd"):
...
+inner = Autograd(Logger(Torch(), name="Torch"), name="Autograd1", create_graph=False)
+outer = Autograd(inner, name="Autograd2", create_graph=False)
+x = label(torch.rand([]), "x")
+
+def run_gradgrad(outer, inner, x):
+    y = square_with_custom_vjp(outer, x)
+    import pdb; pdb.set_trace()
+    dx, = outer.grad(y, [x])
+    # print(outer.gradient_tape)
+    # Bug 1: outer.grad still extend outer.gradient_tape, even though create_graph is False.
```
Comment (zou3519) [inline simple_functorch.py:1307]: Bug 1, which is probably related

File: simple_functorch.py [hunk 4]
```diff
@@ -1277,3 +1277,72 @@ def run_gradvmap(d2: "Batched", d1: "Autograd"):
...
+    ddx, = inner.grad(dx, [x])
+    # Bug 2: inner.gradient_tape
+    # The second TapeEntry doesn't use x at all! In fact, no tapes
+    # capture the 2 * x behavior.
+    # [TapeEntry(inputs=['x'], outputs=['v199'], propagate=<function Autograd.custom_vjp.<locals>
+    # .propagate at 0x7f8310658a60>), TapeEntry(inputs=['v200', 'v198'], outputs=['v201'], propagate
+    # =<function Autograd.mul.<locals>.propagate at 0x7f83106a3dc0>)]
```
Comment (zou3519) [inline simple_functorch.py:1314]: Bug 2
